app/src/main/java/com/example/geomate/ui/screens/groups/GroupsViewModel.kt

```diff
@@ -19,20 +21,52 @@ class GroupsViewModel(
     private val _uiState = MutableStateFlow(GroupsUiState())
     val uiState: StateFlow<GroupsUiState> = _uiState.asStateFlow()
 
+    private val _matchGroups: MutableStateFlow<MutableMap<Group, List<Uri>>> =
+        MutableStateFlow(mutableMapOf())
+    val matchGroups: StateFlow<MutableMap<Group, List<Uri>>> = _matchGroups.asStateFlow()
     fun fetchGroups(userId: String) = viewModelScope.launch {
         groupsRepository.getAll(userId).collect { groups ->
             _uiState.update {
                 it.copy(
                     groups = groups.associateWith { group ->
-                        List(group.users.size) { Uri.EMPTY }
+                        List(group.users.size) {
+                            Uri.EMPTY
+                        }
                     }.toMutableMap()
                 )
             }
+            updateMatchGroups(_uiState.value.groups)
+        }
+    }
+
+    private fun fetchMatchGroups() = viewModelScope.launch {
+        uiState.collect {
+            if (uiState.value.searchQuery.isBlank()) {
+                Firebase.auth.uid?.let { fetchGroups(it) }
+            } else {
+                val groups = matchGroups.value.filterKeys {
+                    it.name.contains(
+                        uiState.value.searchQuery,
+                        true
+                    )
+                }.toMutableMap()
+                updateMatchGroups(groups)
+                updateGroups(groups)
+            }
         }
     }
 
     fun updateSearchQuery(searchQuery: String) {
         _uiState.update { it.copy(searchQuery = searchQuery) }
+        fetchMatchGroups()
```

Comment:
You shouldn't fetch the groups from Firebase on each `searchQuery` change. We have the list of all groups in the uiState, so all you need to do, is to filter them and present only matched groups in the UI layer. The `fetchMatchGroups` method is not needed.

```
fun updateSearchQuery(searchQuery: String) {
        _uiState.update {
              it.copy(
                     searchQuery = searchQuery,
                     matchedGroups = groups.filter { it.name.contains(searchQuery) /* or more complex filtering */ }
              )
       }
}

```

Reply:
Yes, it is exactly like that.